core/modules/date/date.field.inc

```diff
@@ -557,3 +677,442 @@ function date_field_settings_form($field, $instance, $has_data) {
+            )));
+          }
+          else {
+            form_set_error($item_parents . '][value2', t('The end date in %field may not be before @date.', array(
+              '%field' => $field_label,
+              '@date' => $element['value']['#min'],
+            )));
+          }
+        }
+        if (isset($max_date) && $date2 > $max_date) {
+          if ($multi_value) {
+            form_set_error($item_parents . '][value2', t('The end date of value #!delta in %field may not be after @date.', array(
+              '!delta' => $delta + 1,
+              '%field' => $field_label,
+              '@date' => $element['value']['#max'],
+            )));
+          }
+          else {
+            form_set_error($item_parents . '][value2', t('The end date in %field may not be after @date.', array(
+              '%field' => $field_label,
+              '@date' => $element['value']['#max'],
+            )));
+          }
+        }
+        $values[$delta]['offset2'] = $date2->getOffset();
+        $date2->setTimezone(new DateTimeZone($timezone_db));
+        $values[$delta]['value2'] = _date_html5_format_value_db($date2, $field_type);
+        if ($date2 <= $date) {
+          if ($multi_value) {
+            form_set_error($item_parents . '][value2', t('The end date of value #!delta in %field must be greater than the start date.', array(
+              '!delta' => ($delta + 1),
+              '%field' => $field_label,
+            )));
+          }
+          else {
+            form_set_error($item_parents . '][value2', t('The end date in %field must be greater than the start date.', array(
+              '%field' => $field_label,
+            )));
+          }
+        }
+      }
+      else {
+        if (!empty($element['value2']['#required']) && $delta < 1) {
+          if ($multi_value) {
+            form_set_error($item_parents . '][value2', t('An end date for value #1 is required in %field.', array(
+              '%field' => $field_label,
+            )));
+          }
+          else {
+            form_set_error($item_parents . '][value2', t('An end date is required in %field.', array(
+              '%field' => $field_label,
+            )));
+          }
+        }
+        // Function date_field_validate chokes otherwise.
+        $values[$delta]['value2'] = NULL;
+      }
+    }
+
+    // Override the value to what date fields expect.
+    $form_state['values'][$field_name][LANGUAGE_NONE] = $values;
+  }
+}
+
+/**
+ * Prepare values from database for the widget form.
+ *
+ * @param string $value_orig
+ *   Original value from database.
+ * @param string $type
+ *   Form API type of the widget.
+ * @param int $timestep
+ *   Step value for the form, 1 or 60.
+ * @param string $timezone_db
+ *   Timezone of the database value.
+ * @param string $timezone
+ *   Actual timezone for output.
+ *
+ * @return array
+ */
+function _date_html5_get_form_value($value_orig, $type, $timestep, $timezone_db, $timezone) {
```

Comment:
I get the feeling that almost all the private functions in this file should probably be in `date.elements.inc` since that's where other private functions are. That keeps this file cleaner and focused on just defining the field (I think).